File: demo/bridge-demo-adr/app/src/main/java/com/viethai/demo/LocationModuleBridge.kt
package com.viethai.demo

import android.webkit.JavascriptInterface
import android.webkit.WebView
import com.google.gson.Gson

/** Created by viethai.pham on 2019-07-02 */
class LocationModuleBridge(webView: WebView, private val module: LocationModule, gson: Gson) :
  BaseModuleBridge(webView, gson) {

  @JavascriptInterface
  fun observeLocationChange(requestString: String) {
    val request = this.gson.fromJson(requestString, Request::class.java)
    this.sendStreamResponse(request, this.module.observeLocationStream())
  }
}

File: demo/bridge-demo-ios/BridgeDemoIOS/LocationModuleBridge.swift
//
//  LocationModuleBridge.swift
//  BridgeDemoIOS
//
//  Created by Viethai Pham on 2/7/19.
//  Copyright © 2019 Viet Hai Pham. All rights reserved.
//

import WebKit

public final class LocationModuleBridge: BaseModuleBridge {
  private let module: LocationModule
  
  public init(module: LocationModule, delegate: BridgeDelegate) {
    self.module = module
    super.init(delegate)
  }
}

extension LocationModuleBridge: WKScriptMessageHandler {
  public func userContentController(_ userContentController: WKUserContentController,
                                    didReceive message: WKScriptMessage) {
    let params = message.body as! [String : Any]
    let method = params["method"] as! String
    let callback = params["callback"] as! String
    
    switch method {
    case "observeLocationChange":
      let locationStream = self.module.observeLocationChange()
      
      self.sendStreamResult(stream: locationStream, callback: callback)
        .disposed(by: self.disposeBag)
      
    default:
      break
    }
  }
}
